### DISCUSSION TITLE
Quantities via siunitx

### DISCUSSION THREAD

**Discussion by jan-david-fischbach** [jupyter-book/mystmd#539] [Ideas] [answered]
Hey @rowanc1 @stevejpurves @fwkoch,
I usually make heavy use of the `\qty{}{}` command when editing my scientific documents in LaTeX. What would be the best way to support a similar feature (probably role?[^1]) in myst. Would it make sense to extend the mdast spec to include some `type: quantity` which is then expressed using siunitx in .tex output. In the web version we could implement a popover, that allows the interactive conversion to other units or similar.

[^1]: I am not sure, whether roles can support multiple arguments (amount and unit). Furthermore in LaTeX the qty command can also be used in Math environments, which would be nice, to maintain.

**Comment by agoose77** [comment, marked as answer]
I believe that we already have support for siunits in the AST: https://github.com/executablebooks/mystmd/pull/354

The next step is formalising a MyST syntax to parse. @rowanc1 authored that PR, IIRC.

**Reply by jan-david-fischbach** [reply]
Ah ok, didn't see that...

**Reply by rowanc1** [reply]
Yes! We can parse it from latex pretty well, and it does have a representation in the AST, and the UI is there. The MyST role is not yet defined: we need to figure out the MyST syntax, and probably a few other things. If you are interested in doing a bit of work on that I can set up an issue with pointers!

There is an example here (parsed from latex) of a published MyST article that has units in figure 4:
https://journal.physiomeproject.org/articles/S000014#fig-fig4
There is alt-text that shows up over the units! I feel like we can actually do a really good job on accessibility and other things in this -- all of which is lost when getting a printed PDF!
![units](https://github.com/executablebooks/mystmd/assets/913249/0fd58122-edcd-411b-92b3-59e37e058ee1)

**Reply by jan-david-fischbach** [reply]
Jeah, that looks really good! 

I have also thought about simple conversions (eg. inches to metric) but that is rather gimmicky.

I will put some time in to create the MyST syntax, if that is desired. 

Regards JD

**Reply by rowanc1** [reply]
Yeah, I like the idea of those conversions, but at the very least having this be more machine readable and accessible is a huge step on its own!

Awesome, that would be great if you could open an issue suggesting a syntax we can pull in a few more people to think on it after that. Thanks for your help!! 🚀

**Reply by jan-david-fischbach** [reply]
Opened #545

**Comment by scimax** [comment]
Hello everyone, I know it's quite an old discussion, but since the PR #577  was blocked, and as it is at least unclear to me how to use the `si` role (if it exists), I wanted to bring this question up again. So what I understood so far is 
- that `tex-to-myst` can parse the `\qty` and `\unit` commands.
- the PR tries to mimic the `siunitx` formatting behavior, more than just displaying "number" and "unit" more or less in a dumb way. Is that correct? Since this goes beyond the current scope it is blocked, and might make sense for a plugin in the feature.
- In the `myst-to-tex/src/index.ts` there is a `si` role (?), which is also listed in the [reference section of the docs](https://mystmd.org/guide/roles#role-si), but it's unclear to me if it works, and if it does, how it works. I tried something like `` {si}`1 mm` `` 

I would be thankful for support in moving this a bit forward. I could maybe also contribute some dev or docs time, but I'm a bit unfamiliar with typescript which is why I might be a bit slow.

**Reply by jan-david-fischbach** [reply]
To be honest, I am quite lost, what is blocking the PR. To me the lack of proper unit handling is a pretty big annoyance as I am trying to adopt MyST in my scientific writing.

**Reply by stefanv** [reply]
I appreciate that; would it be useful to you, until we sort this out "officially", to provide your directive as a plugin?

**Reply by stefanv** [reply]
Also, I'm trying to dig into this a bit further, and wondering what the use-cases are you're trying to cover: https://github.com/jupyter-book/mystmd/pull/577#issuecomment-4101226486

**Reply by stefanv** [reply]
In terms of a workaround, here's an example of how a plugin could work (this one provides the `si2` directive):

https://gist.github.com/stefanv/00cdc4e815b06668d55cce2ee60fef9c

**Reply by scimax** [reply]
@stefanv I will try to get my hands on your `si2` plugin. It looks like it implements a lot of features of `siunitx` I would like to use. I have to add that I have to get this plugin running, though^^

**Reply by stefanv** [reply]
Here's how to add it to your `myst.yml`:

```yaml
version: 1
project:
  # title:
  # description:
  # keywords: []
  # authors: []
  # github:
  toc:
    - file: index.md
  plugins:
    - si2.mjs
```

This is just an experiment, but if it solves your use case for now that'd be a win :)

**Comment by stefanv** [comment]
I don't know if this is helpful in your case, but I've been using this in my book configuration to handle `\unit`:

```
  math:
    '\unit': '\,\mathrm{#1}'
    '\unitfrac': '\,\tfrac{\mathrm{#1}}{\mathrm{#2}}'
    '\nicefrac': '\tfrac{#1}{#2}'
    '\ointctrclockwiseop': '\oint'
    '\mathparagraph': '\P'
    '\brokenvert': '\big|'
```

Perhaps this workaround is useful in other scenarios too.

**Reply by scimax** [reply]
Thanks, @stefanv , maybe I should rewrite the units using a math macro. I tried it already but had something not matched, and wanted to figure out first, how to do it. 

Of course I used `\per` in the unit command which was not defined. So I should simply try to write the units as plain text then the macro should work.
